### ISSUE TITLE
throw new Error('Must access array elements with a number, not "' + String(first) + '".');

### ISSUE THREAD

**Issue by Xurma2** [erikras/redux-form#3753]
When I make onFocus then onBlur and then again onFocus I get this error

_Must access array elements with a number, not "undefined"_
**"redux-form": "^7.1.1"**

![error](https://user-images.githubusercontent.com/5157416/34911335-a86df1b4-f8d9-11e7-8bc2-af603d2cca07.gif)

**Comment by Xurma2** [comment]
Related https://stackoverflow.com/questions/48014307/reduxform-error-when-changing-value-of-a-radiobutton

**Comment by Xurma2** [comment]
And code 
`            {
              distributions.map((distribution, key) => {
                return <React.Fragment key={key}>
                  <TabPanel>
                    <form name={`aqForm_${distribution.id}`} className={s.form}>
                      <Field
                        key={key}
                        name={"123"}
                        component={TextInput}
                        type={"text"}
                        placeholder={""}
                      />
                    </form>
                  </TabPanel>
                </React.Fragment>;
              })
            }`

I think that related to map
But how it make if I need different form in each tabs?

**Comment by ianwitherow** [comment]
I just ran into this problem and solved it. Within my `map` function for my fields, on the <Field /> name properties, I had some string that made sense to me, but I don't think it was what Redux/Redux Form expected. I updated it to match the format shown in the examples:

```
<Field name={member + ".Name" />
<Field name={member + ".Occupation" />
...
```

Hope this helps!

**Comment by Xurma2** [comment]
@ianwitherow Thanks!
But in my case, this is more of a form.
I need to draw a new <form> for each tab

**Comment by Xurma2** [comment]
`function deleteInWithPath(state, first) {
  if (state === undefined || state === null || first === undefined || first === null) {
    return state;
  }

  if (rest.length) {
    if (Array.isArray(state)) {
      if (isNaN(first)) {
        throw new Error('Must access array elements with a number, not "' + String(first) + '".');
      }
      var firstIndex = Number(first);
      if (firstIndex < state.length) {
        var result = deleteInWithPath.apply(undefined, [state && state[firstIndex]].concat(_toConsumableArray(rest)));
        if (result !== state[firstIndex]) {
          var copy = [].concat(_toConsumableArray(state));
          copy[firstIndex] = result;
          return copy;
        }
      }
      return state;
    }
    if (first in state) {
      var _result = deleteInWithPath.apply(undefined, [state && state[first]].concat(_toConsumableArray(rest)));
      return state[first] === _result ? state : _extends({}, state, _defineProperty({}, first, _result));
    }
    return state;
  }

var deleteIn = function deleteIn(state, field) {
  return deleteInWithPath.apply(undefined, [state].concat(_toConsumableArray((0, _toPath3.default)(field))));
};`

That happened here

**Comment by joyguy55** [comment]
Yeah this error is lingering away in our codebase as well. If anyone could fix/advise why this occurs that would be super helpful. It seems to impact performance. I might consider another library.

**Comment by gustavohenke** [comment]
Reproducible sandbox, anyone?

**Comment by somaholiday** [comment]
I encountered this error too, and it seems to be triggered by naming a field with a number.

Here is a repro case I made by forking the Simple Form example: https://codesandbox.io/s/lr5zwz2mvz

Note focusing/blurring on the `Field` named "name" is fine, but focusing/blurring on the `Field` named "2018" causes the error.

**Comment by davidfurlong** [comment]
Related: https://github.com/erikras/redux-form/issues/2388

**Comment by keeth** [comment]
I ran into this too.

I had a form that was like..
```
<form>
  <Field name='0.firstName'>
  <Field name='0.lastName'>
  <Field name='1.firstName'>
  <Field name='1.lastName'>
</form>
```

The weird thing is that it mostly works, but I get the above error in the console.

The fix (after much head scratching) was to prefix the fields:

```
<form>
  <Field name='guests.0.firstName'>
  <Field name='guests.0.lastName'>
  <Field name='guests.1.firstName'>
  <Field name='guests.1.lastName'>
</form>
```

I dunno, seems like a weird gotcha.. not sure if the best fix is just to clarify in the docs, or what.

**Comment by DarrenBangsund** [comment]
I named my `FormSection` with an id string eg.

```jsx
<form>
  <FormSection name={`${id}`}>
    <Field name={"someFieldString"}/>
  </FormSection>
</form>
```

and was receiving this error as well...changing the interpolated name prop to contain non-number characters solved my issue:

```jsx
<form>
  <FormSection name={`someString${id}`}>
    <Field name={"someFieldString"}/>
  </FormSection>
</form>
```

Bizarre head-scratch-inducing behavior

**Comment by b-fett** [comment]
@gustavohenke, as said @Xurma2, you need to focus twice to get error.
https://codepen.io/anon/pen/eQaWVJ?editors=1011

**Comment by vanamagautam** [comment]
If the type of your input is "file" set your Field name to an empty string and pass value={null} to your input.

**Comment by Youans** [comment]
Redux form states it clear

**Numeric field names, e.g. name="42" or name="foo.5.email", are not supported, as they can be confused for array indexes.**

in https://redux-form.com/7.1.2/docs/api/field.md/#props-you-can-pass-to-code-field-code-

**Comment by dimaqq** [comment]
Sorry to poke an old thread.
I've managed to hit this, and have a Sentry log, but can no longer reproduce...

Here's my analysis:
I have things that are keyed by hex strings, so typically `123ac45` but sometimes `12345678`.
I've been making fields like `items.${ key }`.
Some pair of keys was problematic, and it complained that:
`Must access array elements with a number, not "909a351a".`
As far as I can tell either when the form was unmapped, or when initial values were reinitialised.

Maybe unrelated:

Going through the code, I see a liberal use of `isNaN` to determine if something is a string.
For example at https://github.com/redux-form/redux-form/blob/f8b721b9b33511a7b0be597f0bd665836b70ba8d/src/structure/plain/deleteIn.js#L12-L16

This test is however problematic in itself, because:
```
isNaN("33")
false
isNaN("33a")
true
isNaN("1e1")
false
```

That is, non-numeric "path" element is mistakenly assumed to be numeric if it happens to form valid scientific notation.

**Comment by jdvivar-eb** [comment]
We discovered the root cause was that `redux-form` created a `fields` value as an array in the redux store if all field names were numbers, whereas if only one field was not a number, it worked.

So we fixed this issue by adding an ephemeral field only for the purpose of creating a valid `fields` object (rather than an array):
```js
dispatch(touch(FORM_NAME, 'ephemeral-field'));
```

This forced redux-form to create a `fields` object. Later actions performed by number-only fields worked correctly.

**Comment by dimaqq** [comment]
This library cannot always guess for if array or object shape is intended (it does guess, but sometimes the data is misleading).

The standard (?) work-around is to prefix the keys when developer expects an object, and be explicit (in code) when developer expects and array, thus:
* `"id-938588282", "id-091939392"` instead of `938588282", 091939392`
* `0, 1, 2, 3` instead of `938588282", 091939392, ...`
